File: Day79/Part1/time_frozen_enemies.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Input format:
// n T D
// spawn[0..n-1]
// hp[0..n-1]
// Output: maximum number of enemies killed during a single freeze

long long solve() {
    int n; long long T, D;
    if (!(cin >> n >> T >> D)) return 0;
    vector<long long> spawn(n), hp(n);
    for (int i = 0; i < n; ++i) cin >> spawn[i];
    for (int i = 0; i < n; ++i) cin >> hp[i];

    long long capacity = T * D; // total damage we can deal during freeze

    // We may activate freeze at any time; best is after all are spawned.
    // Choose the smallest-HP enemies to maximize count within capacity.
    sort(hp.begin(), hp.end());

    long long sum = 0; long long cnt = 0;
    for (long long h : hp) {
        if (sum + h <= capacity) {
            sum += h; cnt++;
        } else break;
    }
    return cnt;
}

int main() {
    cout << solve() << endl;
    return 0;
}

File: Day79/Part1/time_frozen_enemies.py
import sys

def solve():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    it = iter(data)
    try:
        n = int(next(it)); T = int(next(it)); D = int(next(it))
    except StopIteration:
        return
    spawn = []
    hp = []
    for _ in range(n):
        try:
            spawn.append(int(next(it)))
        except StopIteration:
            spawn.append(0)
    for _ in range(n):
        try:
            hp.append(int(next(it)))
        except StopIteration:
            hp.append(0)

    capacity = T * D
    hp.sort()
    s = 0
    cnt = 0
    for h in hp:
        if s + h <= capacity:
            s += h
            cnt += 1
        else:
            break
    print(cnt)

if __name__ == "__main__":
    solve()
